api: Add an endpoint for listing the user groups in realm.

zy964c opened the pull request:
Adds endpoint for listing the user groups in realm.
Provides "description", "members name" and "id" for each user group.

Fixes #10298

YagoGG commented on zerver/tests/test_user_groups.py:
```diff
@@ -110,6 +110,13 @@ def test_user_group_create(self) -> None:
+        self.login(self.example_email("hamlet"))
+        result = self.client_get('/json/user_groups')
+        self.assert_json_success(result)
+        self.assert_length(UserGroup.objects.all(), 1)
```
I'm not sure what you're trying to test here. Maybe it should be something like:

```py
# ... Get Hamlet's UserProfile as user_profile
self.assert_length(result.json()['user_groups'], UserGroup.objects.filter(realm=user_profile.realm).count())
```
Note that we should only filter the user groups in this realm, since that's what the endpoint is actually doing.

YagoGG commented:
Hey @zy964c, thanks for working on this!

I have been able to test the feature and it works nice. I just left a comment regarding the tests, once that's fixed this should be ready to go.

Cheers!

zy964c commented:
Hi @YagoGG , I've made that change test change.
Thanks!

YagoGG commented on zerver/tests/test_user_groups.py:
```diff
@@ -110,6 +110,14 @@ def test_user_group_create(self) -> None:
+    def test_user_group_get(self) -> None:
+        # Test success
+        user_profile = self.example_user('hamlet')
+        self.login(self.example_email("hamlet"))
```
You can grab the email here with `user_profile.email`.

zy964c replied:
OK, have changed that.

zy964c commented:
@YagoGG Yep, I'll add the method and update `.yaml` file later today

zy964c commented:
Hey @YagoGG , I've done the following on the follow ups:
* Created the corresponding method in api python bindings: zulip/python-zulip-api#474
* Edited `zulip.yaml` via 7f28276. I'm not sure if that is all changes required  to document new endpoint.

Please let me know if any corrections/additions needed.
Thanks!

YagoGG commented on zerver/openapi/zulip.yaml:
```diff
@@ -1048,6 +1048,59 @@ paths:
+                        `description`, a `name`, their `id` and the list of members
+                        of the user group.
+                - example:
+                    {
```
Even though the YAML file is indented with 2 spaces, we use 4 spaces for JSON responses to make them more readable.

zy964c replied:
@YagoGG , I added spaces for responses and removed `400:` fixture.

YagoGG commented on zerver/openapi/zulip.yaml:
```diff
@@ -1048,6 +1048,59 @@ paths:
+                      }
+                      ]
+                    }
+        '400':
```
We usually add error responses if they are specific for an endpoint or they may cause confusion. This one is pretty clear in my opinion, so I think we can remove this response.

YagoGG commented:
This is looking great, @zy964c! I just left a couple comments on the YAML changes.

Usually, after adding an endpoint to the OpenAPI file, we also have a matching article in our API docs. You can see an example in #10343, if you feel like adding it as well.

YagoGG commented on templates/zerver/api/get-user-groups.md:
```diff
@@ -0,0 +1,52 @@
+## Response
+
+#### Return values
+
```
This should have a parent `user_groups` bullet, with all of these descriptions under it as children elements.

zy964c replied:
@YagoGG I added this bullet.

YagoGG commented on templates/zerver/api/get-user-groups.md:
```diff
@@ -0,0 +1,52 @@
+
+## Arguments
+
+This endpoint takes no arguments.
```
You can replace this with:

    {generate_api_arguments_table|zulip.yaml|/user_groups:get}

And it will automatically point out that the endpoint doesn't take arguments.

zy964c replied:
@YagoGG Thanks! Replaced it.

YagoGG commented:
@zy964c Good job with the documentation article! Left a couple notes.

Don't worry about the JS bindings, they don't cover this endpoint and that's fine for now.

There are two things left to give this the final touch:

- Add an entry to `templates/zerver/help/include/rest-endpoints.md`, pointing to the new article. (That's how you'll be able to see it in the docs' TOC and check that everything looks good).
- Include the new endpoint in the API tests, which can be found in `zerver/lib/api_test_helpers.py`.

This is an example of how should it all look like: https://github.com/zulip/zulip/commit/aa5185fdf8cf144ec556cd69184e126c00f871a4. Let me know if you need any help along the process.

zy964c commented:
@YagoGG Thanks for the review. I added entry to TOC and updated `zerver/lib/api_test_helpers.py`.
Now I can not pass api test because, in my understanding, zulip/python-zulip-api#474 should be merged first for Client to have `get_user_groups` method

YagoGG commented on zerver/lib/api_test_helpers.py:
```diff
@@ -204,6 +204,21 @@ def get_streams(client):
+    assert user_groups[0]['description'] == 'Other users'
+
+    validate_against_openapi_schema(result, '/user_groups', 'get', '200')
+    assert len(result['streams']) == 2
```
I guess these two lines popped up after a copy-paste and shouldn't be here anymore.

YagoGG commented:
Yeah, the tests won't pass until that is merged and we bump the server's version of `python-zulip-api/zulip`. Remember to add a commit before all of your work doing the bump, once we're done with https://github.com/zulip/python-zulip-api/pull/474. @timabbott @eeshangarg Can you take a look whenever you have a few minutes?

I left a small comment, but besides that I think this should be ready to merge. Great job, @zy964c!

timabbott commented:
I merged the first commit; let me know when the API stuff means this is ready to merge.

zy964c commented:
@YagoGG to bump server's version of `python-zulip-api/zulip` should some files in `zulip/requirements` be updated, right?

timabbott commented:
Yep, https://zulip.readthedocs.io/en/latest/subsystems/dependencies.html should cover what you need to know (and `git log -p requirements/` might help too)

YagoGG commented:
@zy964c I just bumped the package's version myself in #10344, so you can wait for it to be merged and then simply rebase, if you prefer.

zy964c commented:
Thanks @YagoGG I have rebased last two commits. @timabbott this might be ready for merge
